File: controllers/characterController.js
const characterService = require('../services/characterService.js');

exports.searchCharacters = async (req, res) => {
  try {
    const { name, server, dataCenter } = req.query;
    
    if (!name) {
      return res.status(400).json({
        success: false,
        message: 'Le nom du personnage est requis'
      });
    }

    const capitalizeFirstLetter = (str) => 
      str?.charAt(0).toUpperCase() + str?.slice(1).toLowerCase();
    
    const characters = await characterService.searchCharacters(name, capitalizeFirstLetter(server), capitalizeFirstLetter(dataCenter));
    
    res.status(200).json({
      success: true,
      numberOfCharacters: characters.length,
      characters: characters.sort((a, b) => a.name.localeCompare(b.name)) // tri des personnages par nom alphabétique
    });
  } catch (error) {
    console.error('Erreur lors de la recherche des personnages:', error);
    res.status(500).json({
      success: false,
      message: 'Erreur lors de la recherche des personnages',
      error: error.message
    });
  }
};

exports.getCharacterDetails = async (req, res) => {
  try {
    const { id } = req.params;
    
    if (!id) {
      return res.status(400).json({
        success: false,
        message: 'L\'ID du personnage est requis'
      });
    }
    
    const character = await characterService.getCharacterDetails(id);
    
    res.status(200).json({
      success: true,
      character: character
    });
  } catch (error) {
    console.error('Erreur lors de la récupération des détails du personnage:', error);
    res.status(500).json({
      success: false,
      message: 'Erreur lors de la récupération des détails du personnage',
      error: error.message
    });
  }
};

File: controllers/character_controller.py
import asyncio
from aiohttp import ClientResponseError
from flask import Response, request, jsonify
from http import HTTPStatus
from requests import HTTPError
from services.character_service import (
    search_characters_service,
    get_character_details_service,
    get_character_achievements_service,
)


def search_characters() -> tuple[Response, HTTPStatus]:
    """
    Search for characters based on name and optionally server
    """
    try:
        name = request.args.get("name")
        server = request.args.get("server", "").capitalize()

        if not name:
            return (
                jsonify({"success": False, "message": "Character's name is required"}),
                HTTPStatus.BAD_REQUEST,
            )

        characters = search_characters_service(name, server)

        return (
            jsonify(
                {
                    "success": True,
                    "numberOfCharacters": len(characters),
                    "characters": (
                        sorted(characters, key=lambda char: char["name"])
                        if len(characters) > 0
                        else "No character found"
                    ),
                }
            ),
            HTTPStatus.OK,
        )

    # Serveur incorrect
    except ValueError as error:
        return (
            jsonify({"success": False, "message": str(error)}),
            HTTPStatus.BAD_REQUEST,
        )

    except Exception as error:
        return (
            jsonify(
                {
                    "success": False,
                    "message": "Erreur while searching characters",
                    "error": str(error),
                }
            ),
            HTTPStatus.INTERNAL_SERVER_ERROR,
        )


def get_character_details(id: int) -> tuple[Response, HTTPStatus]:
    """
    Retrieves a character's details by ID.
    """
    try:
        character = get_character_details_service(id)

        return jsonify({"success": True, "character": character}), HTTPStatus.OK

    except HTTPError as error:
        return (
            jsonify({"success": False, "message": "Character not found"}),
            HTTPStatus.NOT_FOUND,
        )

    except Exception as error:
        return (
            jsonify(
                {
                    "success": False,
                    "message": "Erreur while getting character's data",
                    "error": str(error),
                }
            ),
            HTTPStatus.INTERNAL_SERVER_ERROR,
        )


def get_character_achievements(id: int) -> tuple[Response, HTTPStatus]:
    """
    Search character achievements by ID
    """
    try:
        achievements = asyncio.run(get_character_achievements_service(id))

        return (
            jsonify(
                {"success": True, "character_id": id, "achievements": achievements}
            ),
            HTTPStatus.OK,
        )

    except ClientResponseError as error:
        if error.status == HTTPStatus.FORBIDDEN:
            message = "This character has their achievements private"
            status = HTTPStatus.FORBIDDEN
        else:
            message = "Character not found"
            status = HTTPStatus.NOT_FOUND

        return jsonify({"success": False, "message": message}), status

    except Exception as error:
        return (
            jsonify(
                {
                    "success": False,
                    "message": "Erreur while getting character's achivements",
                    "error": str(error),
                }
            ),
            HTTPStatus.INTERNAL_SERVER_ERROR,
        )
